### PR TITLE
Altp2m support

### PR DESCRIPTION
**Pull request by K-Mayer**
This adds support for the direct control of the altp2m-features offered by Xen.

I don`t know if exposing these features to a user without any sanity checks fits in with the spirit of libvmi.
But since I use them constantly I hope they can be usefull for others.

Any comments / suggestions?

Cheers

Maro

### CODE REVIEW

**Comment by CMF-Marodeur** [comment]
Just saw an error when declaring / implementing the altp2m-functions.
Were should the different functions be declared / implemented?
Maybe like this:
vmi_altp2m_get_domain_state -> accessors + libvmi.h
vmi_altp2m_set_domain_state -> accessors + libvmi.h
vmi_altp2m_create_p2m -> ?
vmi_altp2m_destroy_p2m -> ?
vmi_altp2m_switch_p2m -> ?
vmi_altp2m_set_mem_access -> events
vmi_altp2m_change_gfn -> accessors + libvmi.h

**Comment by tklengyel** [comment]
So I had been debating about adding this wrapper to LibVMI as well. The thing to consider here is that there is no support for altp2m-type setup in KVM or any other hypervisor but Xen. As such, adding this wrapper which simply forwards to libxc 1:1 adds little value. So I'm torn whether we should have this added to the LibVMI API or not at this time.

@smaresca @awsaba any opinion?

**Comment by smaresca** [comment]
Even if no other hypervisor implements a feature, I think it's likely fine to expose it so other future drivers can make use of it.

**Comment by tklengyel** [comment]
@smaresca Sure. In that case I would prefer a hypervisor-neutral naming for it though. We already use VMM Pagetable ID instead of altp2m_idx for this reason in the events API. Maybe vmi_vmmpt_*?

**Comment by smaresca** [comment]
I feel like the expression of such a capability in the API will need a more generic name though

**Comment by tklengyel** [comment]
This is ok to test.

**Comment by smaresca** [comment]
@CMF-Marodeur could you describe your current use-case? We're definitely familiar with alt-p2m and its purpose, but knowing how you make use of it could help shape the API.

**Comment by awsaba** [comment]
I'd say include it, but just barely, because the EPTP is architectural feature that may or may not be supported by the hardware itself.  If included, it should get it's own header,  vmi_intel (or vmi_x86, does SVM have any equivalent?) because being generic is nice, but having libvmi keep track of an xch and domid is definitely helpful.

**Comment by tklengyel** [comment]
@awsaba: it's not necessarily architectural. We are actually adding altp2m support to Xen on ARM right now. But the name altp2m is definitely Xen-specific.

**Comment by tklengyel** [comment]
@CMF-Marodeur: to answer your question, I think this could go into a separate header that get's installed with the other libvmi headers. It doesn't really belong to the core libvmi API or the events API. The one altp2m function (altp2m_set_mem_access) that belongs to the events API already is already exposed via the standard LibVMI events API.

**Comment by CMF-Marodeur** [comment]
@smaresca I use the altp2m-feature to quickly switch between two sets of access rights. Basically I divide the guest memory into OS-specific memory and third-party memory because I`m interestested in the transition between those memory regions.
Something a bit like CXPInspector by Carsten Willems et al..

In my use case I also discarded the libvmi event handling (to much overhead since I have dozens / hundreds of memory pages which should all call the same callback) and instead I just defined a default handler for write and one for execute events. For this I implemented another set_mem_access which does not rely on the libvmi events:
status_t vmi_altp2m_set_mem_access ( vmi_instance_t vmi , uint16_t view, vmi_mem_access_t access, addr_t gfn )
Problem ist that this approach is a pretty dirty hack (in the event handling) and would require quite some work for a clean and configurable implementation.

Name:
I assume something containing EPT is discarded because it is not vendor agnostic?

**Comment by tklengyel** [comment]
@CMF-Marodeur: indeed, the current event API has some overhead in that it requires separate event structures for each page it is registered on. We could probably make it more flexible to have a subset of the mem_access event type that registers to a type of violation rather then specific gfns.

**Comment by CMF-Marodeur** [comment]
@tklengyel : That is exactly what I needed (in combination with a way to set the access rights without having to create an event struct first).

Concerning this PR:
So I basically create a new header which exposes the functions. The implementation stays in xen_events.c.
Since you guys already use VMM Pagetable ID i could stick with vmi_vmmpt_*. For example

vmi_vmmpt_create
vmi_vmmpt_switch
vmi_vmmpt_switch

As an alternative I would suggest vmi_slat_*.
[slat](https://en.wikipedia.org/wiki/Second_Level_Address_Translation)

**Comment by tklengyel** [comment]
Sure, let's go with slat, that is at least standard naming.

**Comment by tklengyel** [comment]
And as for naming within the xen-driver, it should stick to altp2m, only in the public-header and driver interface should it follow the hypervisor neutral-naming.

**Comment by smaresca** [comment]
+1

**Comment by CMF-Marodeur** [comment]
So I moved the slat-code to a dedicated header / source file and changed the names.
Additionally I extendet the makefile etc. so programs using libvmi can access the new functions.

And right after the commit I saw two minor typos / wrong text messages. ^^

File: libvmi/Makefile.am [hunk 1]
```diff
@@ -2,6 +2,7 @@ SUBDIRS = config
 
 h_public =  libvmi.h \
             libvmi_extra.h \
+			slat.h \
```
Comment (tklengyel) [inline libvmi/Makefile.am]: alignment doesn't match, are you using tabs instead of spaces?

File: libvmi/Makefile.am [hunk 2]
```diff
@@ -28,6 +29,7 @@ c_sources = \
     performance.c \
     pretty_print.c \
     read.c \
+	slat.c \
```
Comment (tklengyel) [inline libvmi/Makefile.am]: alignment

File: libvmi/accessors.c [hunk 3]
```diff
@@ -511,4 +511,4 @@ vmi_get_linux_sysmap(
 
     return linux_instance->sysmap;
 
-}
+}
```
Comment (tklengyel) [inline libvmi/accessors.c]: Unnecessary change.

File: libvmi/driver/driver_interface.h [hunk 4]
```diff
@@ -138,6 +138,26 @@ typedef struct driver_interface {
     status_t (*set_debug_event_ptr)(
         vmi_instance_t,
         bool enabled);
+    status_t ( *slat_get_domain_state_ptr )(
```
Comment (tklengyel) [inline libvmi/driver/driver_interface.h]: Style doesn't match the others here (ie. no spaces before and after the function name)

File: libvmi/events.c [hunk 5]
```diff
@@ -754,4 +754,4 @@ status_t vmi_shutdown_single_step(vmi_instance_t vmi)
     {
         return VMI_FAILURE;
     }
-}
+}
```
Comment (tklengyel) [inline libvmi/events.c]: Unnecessary change.

File: libvmi/slat.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+* You should have received a copy of the GNU Lesser General Public License
+* along with LibVMI.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "private.h"
+#include "driver/driver_wrapper.h"
+
+status_t vmi_slat_get_domain_state ( vmi_instance_t vmi , bool *state )
+{
+    status_t rc = VMI_FAILURE;
```
Comment (tklengyel) [inline libvmi/slat.c]: No need for defining status_t here, just return the call to the driver directly.

File: libvmi/slat.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
+/* The LibVMI Library is an introspection library that simplifies access to
+* memory in a target virtual machine or in a file containing a dump of
+* a system's physical memory.  LibVMI is based on the XenAccess Library.
+*
+* Copyright 2011 Sandia Corporation. Under the terms of Contract
+* DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government
+* retains certain rights in this software.
+*
+* Author:
```
Comment (tklengyel) [inline libvmi/slat.h]: Author missing.

File: libvmi/slat.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
+/* The LibVMI Library is an introspection library that simplifies access to
+* memory in a target virtual machine or in a file containing a dump of
+* a system's physical memory.  LibVMI is based on the XenAccess Library.
+*
+* Copyright 2011 Sandia Corporation. Under the terms of Contract
```
Comment (tklengyel) [inline libvmi/slat.h]: I don't think you are under contract with Sandia, so remove this block from here.

File: libvmi/slat.c [hunk 9]
```diff
@@ -0,0 +1,70 @@
+/* The LibVMI Library is an introspection library that simplifies access to
+* memory in a target virtual machine or in a file containing a dump of
+* a system's physical memory.  LibVMI is based on the XenAccess Library.
+*
+* Copyright 2011 Sandia Corporation. Under the terms of Contract
```
Comment (tklengyel) [inline libvmi/slat.c]: Not under Sandia contract, no need to add this.

File: libvmi/slat.c [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+* DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government
+* retains certain rights in this software.
+*
+* Author:
```
Comment (tklengyel) [inline libvmi/slat.c]: Author missing.

File: libvmi/slat.h [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with LibVMI.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/**
+* @file slat.h
+* @brief The LibVMI second level address translation API is defined here.
+*
+* More detailed description can go here.
```
Comment (tklengyel) [inline libvmi/slat.h]: More detailed description can go here. ;)

File: libvmi/driver/xen/xen_events.c [hunk 12]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
     return vrc;
 }
+
+status_t xen_altp2m_get_domain_state ( vmi_instance_t vmi , bool *state )
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    unsigned long domain_id = xen_get_domainid ( vmi );
+    printf ( "in xen_altp2m_get_domain_state with domain_id: %lx\n" , domain_id );
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: Debug message should be printed with dbprint().

File: libvmi/driver/xen/xen_events.c [hunk 13]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
 
     return vrc;
 }
+
+status_t xen_altp2m_get_domain_state ( vmi_instance_t vmi , bool *state )
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    unsigned long domain_id = xen_get_domainid ( vmi );
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: This should be domid_t

File: libvmi/driver/xen/xen_events.c [hunk 14]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    printf ( "in xen_altp2m_get_domain_state with domain_id: %lx\n" , domain_id );
+    if ( !xch ) {
+        errprint ( "%s error: invalid xc_interface handle\n" , __FUNCTION__ );
+        return VMI_FAILURE;
+    }
+    if ( domain_id == VMI_INVALID_DOMID ) {
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: needs casting (ie. (domid)VMI_INVALID_DOMID ).

File: libvmi/driver/xen/xen_events.c [hunk 15]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    return VMI_SUCCESS;
+
+}
+
+status_t xen_altp2m_set_domain_state ( vmi_instance_t vmi , bool state )
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    unsigned long domain_id = xen_get_domainid ( vmi );
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: This should be domid_t

File: libvmi/driver/xen/xen_events.c [hunk 16]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    if ( !xch ) {
+        errprint ( "%s error: invalid xc_interface handle\n" , __FUNCTION__ );
+        return VMI_FAILURE;
+    }
+    if ( domain_id == VMI_INVALID_DOMID ) {
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: needs casting (ie. (domid_t)VMI_INVALID_DOMID ).

File: libvmi/driver/xen/xen_events.c [hunk 17]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    domid_t domain_id = xen_get_domainid ( vmi );
+    if ( !xch ) {
+        errprint ( "%s error: invalid xc_interface handle\n" , __FUNCTION__ );
+        return VMI_FAILURE;
+    }
+    if ( domain_id == VMI_INVALID_DOMID ) {
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: needs casting (ie. (domid_t)VMI_INVALID_DOMID ).

File: libvmi/driver/xen/xen_events.c [hunk 18]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    return VMI_SUCCESS;
+
+}
+
+status_t xen_altp2m_create_p2m ( vmi_instance_t vmi , uint16_t *altp2m_idx )
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    domid_t domain_id = xen_get_domainid ( vmi );
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: This should be domid_t

File: libvmi/driver/xen/xen_events.c [hunk 19]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    return VMI_SUCCESS;
+
+}
+
+status_t xen_altp2m_destroy_p2m ( vmi_instance_t vmi , uint16_t altp2m_idx )
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    domid_t domain_id = xen_get_domainid ( vmi );
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: This should be domid_t

File: libvmi/driver/xen/xen_events.c [hunk 20]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    if ( !xch ) {
+        errprint ( "%s error: invalid xc_interface handle\n" , __FUNCTION__ );
+        return VMI_FAILURE;
+    }
+    if ( domain_id == VMI_INVALID_DOMID ) {
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: casting

File: libvmi/driver/xen/xen_events.c [hunk 21]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
...
+    int rc;
+    xen_instance_t *xen = xen_get_instance ( vmi );
+    xc_interface * xch = xen_get_xchandle ( vmi );
+    domid_t domain_id = xen_get_domainid ( vmi );
+
+    if ( !xch ) {
+        errprint ( "%s error: invalid xc_interface handle\n" , __FUNCTION__ );
+        return VMI_FAILURE;
+    }
+    if ( domain_id == VMI_INVALID_DOMID ) {
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: casting missing

File: libvmi/driver/xen/xen_events.c [hunk 22]
```diff
@@ -1410,3 +1417,147 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
 
     return vrc;
 }
+
+status_t xen_altp2m_get_domain_state ( vmi_instance_t vmi , bool *state )
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: These functions should be in a separate file in the Xen folder, say altp2m.c, as these don't belong to the events system.

File: libvmi/driver/xen/xen_events.c [hunk 23]
```diff
@@ -728,6 +728,13 @@ status_t xen_init_events_new(vmi_instance_t vmi)
     vmi->driver.set_cpuid_event_ptr = &xen_set_cpuid_event;
     vmi->driver.set_debug_event_ptr = &xen_set_debug_event;
 
+    vmi->driver.slat_get_domain_state_ptr = &xen_altp2m_get_domain_state;
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events.c]: Wiring up these functions should be done in xen.c and only if the Xen version is 4.6 or newer.

File: libvmi/driver/xen/xen_events_private.h [hunk 24]
```diff
@@ -144,6 +144,15 @@ status_t xen_set_cpuid_event(vmi_instance_t vmi, bool enabled);
 status_t xen_start_single_step(vmi_instance_t vmi, single_step_event_t *event);
 status_t xen_stop_single_step(vmi_instance_t vmi, uint32_t vcpu);
 status_t xen_shutdown_single_step(vmi_instance_t vmi);
+status_t xen_altp2m_get_domain_state (vmi_instance_t vmi, bool *state);
```
Comment (tklengyel) [inline libvmi/driver/xen/xen_events_private.h]: Should be moved to another header, altp2m_private.h which gets included in xen.c

File: libvmi/Makefile.am (hunk 1 above)
Comment (CMF-Marodeur) [inline libvmi/Makefile.am]: Actually yes (unintentionally).
My IDE seems to be forgetting that it should use 4 spaces instead of tabs.

File: libvmi/slat.c (hunk 10 above)
Comment (CMF-Marodeur) [inline libvmi/slat.c]: How much information about the author do you need?
Is a name (real name) sufficent?

File: libvmi/slat.h (hunk 11 above)
Comment (CMF-Marodeur) [inline libvmi/slat.h]: I saw this in the libvmi.h (and other headers), so I thought it would be ok for now. ^^
I will try to think of a more detailed description.

File: libvmi/driver/xen/xen_events_private.h (hunk 24 above)
Comment (CMF-Marodeur) [inline libvmi/driver/xen/xen_events_private.h]: Does the new header require the copyright/licences like the xen_events_private.h
[...]

> * Portions of this header and dependent code is based upon that in xen-access,
> *    from the official Xen source distribution.  That code carries the
> *    following copyright notices and license.
> *
> * Copyright (c) 2011 Virtuata, Inc.
> * Copyright (c) 2009 by Citrix Systems, Inc. (Patrick Colp), based on
> *   xenpaging.c
[...]

**Comment by CMF-Marodeur** [comment]
Baby steps, but I think we are getting there.

Author is still intentionally left blank.
Do you guys need a real name for that?

Concerning the xen_init_altp2m in xen.c:
Technically it can not fail so the return status_t is of no importance there. Not sure if it should stay there or not.

File: libvmi/accessors.c [hunk 25]
```diff
@@ -495,16 +495,16 @@ addr_t vmi_translate_uv2p (vmi_instance_t vmi, addr_t virt_address, vmi_pid_t pi
 
 const char *
 vmi_get_linux_sysmap(
-	vmi_instance_t vmi)
+    vmi_instance_t vmi)
 {
```
Comment (CMF-Marodeur) [inline libvmi/accessors.c]: unrelated fixes (tabs were used there). Should they stay in?

File: libvmi/driver/xen/xen.c [hunk 26]
```diff
@@ -835,6 +836,26 @@ xen_init_events(
 }
 
 status_t
+xen_init_altp2m (
```
Comment (tklengyel) [inline libvmi/driver/xen/xen.c]: So since there is no separate flag the user sends to specifically activate the altp2m parts this function doesn't need to return a status_t and can just be void. You could also make it into a static inline and move it to altp2m_private.h

File: libvmi/accessors.c (hunk 25 above)
Comment (tklengyel) [inline libvmi/accessors.c]: Fine by me.

File: libvmi/Makefile.am [hunk 27]
```diff
@@ -66,7 +68,8 @@ endif
 
 if HAVE_XEN
 h_public    += events.h
-drivers     += driver/xen/xen.h \
+drivers     += driver/xen/altp2m.c \
```
Comment (tklengyel) [inline libvmi/Makefile.am]: Also add altp2m_private.h here.

**Comment by tklengyel** [comment]
This is shaping up fine so with the last two things fixed +1 from me.

File: libvmi/driver/xen/xen.c (hunk 26 above)
Comment (CMF-Marodeur) [inline libvmi/driver/xen/xen.c]: I followed the given scheme, because I think that maybe there should be a flag there.
But switching it to void is fine with me. Any user will see the dbprint from driver_wrapper.h if he tries to use the functions with the wrong xen-version.

**Comment by CMF-Marodeur** [comment]
What about the Author?
Is CMF-Marodeur (no email) fine with you guys?

**Comment by K-Mayer** [comment]
Any comments?

**Comment by tklengyel** [comment]
We will need another +1 from a maintainer and then we can merge. PRs that add/change the API need two maintainers to ack it.

**Comment by awsaba** [comment]
Needs a squash, but otherwise LGTM.

@tklengyel should #382 be merged first and this PR rebased to prevent new warnings from popping up?

**Comment by tklengyel** [comment]
@awsaba: either way works for me

**Comment by smaresca** [comment]
@K-Mayer thanks for the squash -- merging now
